kiwmi/desktop/xdg_shell.c

```diff
@@ -22,6 +23,123 @@
 #include "input/seat.h"
 #include "server.h"
 
+static struct kiwmi_view_child *view_child_popup_create(
+    struct kiwmi_view_child *parent,
+    struct kiwmi_view *view,
+    struct wlr_xdg_popup *wlr_popup);
+
+static void
+popup_new_popup_notify(struct wl_listener *listener, void *data)
+{
+    struct kiwmi_view_child *popup =
+        wl_container_of(listener, popup, new_popup);
+    struct wlr_xdg_popup *wlr_popup = data;
+    view_child_popup_create(popup, popup->view, wlr_popup);
+}
+
+static void
+popup_owner_destroy_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *popup =
+        wl_container_of(listener, popup, owner_destroy);
+    view_child_destroy(popup);
+}
+
+static void
+popup_unconstrain(struct kiwmi_view_child *popup)
+{
+    if (popup->type != KIWMI_VIEW_CHILD_XDG_POPUP) {
+        wlr_log(WLR_ERROR, "Expected an xdg_popup kiwmi_view_child");
+        return;
+    }
+
+    struct kiwmi_view *view         = popup->view;
+    struct wlr_xdg_popup *wlr_popup = popup->wlr_xdg_popup;
+
+    // Q: are the coord calculations across this function correct? I have
```

Comment:
This is how I see it now, with help from Sway’s [`subsurface_get_root_coords`](https://github.com/swaywm/sway/blob/bb3fd0abc56f39e35dc7f4e86f25da1b4a6efbd7/sway/tree/view.c#L904)

```
,-- ly=0
|   (nothing from the view is displayed here)
| . . vy == -view->geom.y; toplevel surface begin
| .   (e.g. drop-shadow)
| . ,-- vy=0; ly == view->y; "user-visible" view area begin
| . |   (this is the actual window)
| . | ,-- sy=0; vy == surf->y; subsurface begin (sy and surf refer to the subsurface here)
| . | |   (here is the subsurface)
```

which also seems to match our rendering (note that it has `output_ly` negative)

```c
int oy = rdata->output_ly + sy + view->y - view->geom.y;
```

(just writing it here because I realised that one of their commits I wanted to look at introduced `subsurface_get_root_coords` and this was the only reason for rechecking it I could think of)

Reply:
I’m quite sure now that it is correct here. Feel free to reopen this if you see it differently.